Bubblegum tests: baseline redeem, cancel, and decompress

alexandruag opened the pull request:
This PR is created (for now) on top of https://github.com/metaplex-foundation/metaplex-program-library/pull/752, so the first four commits are duplicated. It introduces the baseline tests for (cancel) redeem, together with decompress. A few checks are omitted for now due to some weird behaviour that needs to be sorted out (will update the PR afterwards).

alexandruag commented:
#752 got merged a few days ago, and this finally got rebased and in better shape for review.

alexandruag commented on bubblegum/program/build.rs:
```diff
@@ -11,20 +11,27 @@ fn main() {
         Path::new("../../target/deploy/GRoLLzvxpxxu2PGNJMMeZPyMxjAUH9pKqxGXV9DGiceU.so");
     let spl_wrapper_so_path =
         Path::new("../../target/deploy/WRAPYChf58WFCnyjXKJHtrPgzKXgHp6MD9aVDqJBbGh.so");
+    let mpl_tm_so_path = Path::new("../../target/deploy/mpl_token_metadata.so");
```
Currently the `build.rs` scripts in Rust have no (straightforward) way of knowing whether it's a test build or not, so this runs on every build, but nothing really happens if that path already exists. If we ever need/want to only run stuff like this for tests, there are ways to orchestrate that, but they're more convoluted (and generally involve breaking things into multiple crates, with test dependencies).

danenbm replied:
I think we just need a linear task (not needing to be resolved in this PR) to fix the issue where the build.rs script prevents our automation from publishing to crates.io.

alexandruag commented on bubblegum/program/build-token-metadata.sh:
```diff
@@ -0,0 +1,13 @@
+set -e
+
+cd ../../token-metadata
+cargo build-bpf
```
oh that's totally right ... will incorporate

alexandruag commented:
Rebased on top of latest upstream.

danenbm commented on bubblegum/program/tests/simple.rs:
```diff
@@ -54,54 +65,211 @@ async fn test_create_tree_and_mint_passes() {
+        // Test master edition account.
+        let me_key = find_master_edition_account(&mint_key).0;
+        let me_account = tree.read_account(me_key).await.unwrap();
+        let me: MasterEditionV2 =
+            MasterEditionV2::safe_deserialize(me_account.data.as_slice()).unwrap();
```
Similar to the Metadata I believe you could use `MasterEditionV2::from_account_info()` and get one extra check.

danenbm commented on bubblegum/program/tests/simple.rs:
```diff
@@ -54,54 +65,211 @@ async fn test_create_tree_and_mint_passes() {
+        assert!(!meta.primary_sale_happened);
+        assert!(!meta.is_mutable);
+
+        // TODO: Is the edition nonce field predictable so we can/should test for a specific
+        // value there as well?
```
I can look at this tomorrow or you can ask @samuelvanderwaal he would probably know off the top of his head.

danenbm commented on bubblegum/program/tests/simple.rs:
```diff
@@ -54,54 +65,211 @@ async fn test_create_tree_and_mint_passes() {
+        //     assert_eq!(c1.share, c2.share);
+        // }
+
+        assert!(!meta.primary_sale_happened);
```
I am assuming you are testing pretty much every field for these account data types.  I don't think I need to double check your coverage but lmk if I should.

alexandruag replied:
I've switched to an approach where I instantiate the entire expected struct such that we are sure no fields etc. are missing.

danenbm replied:
Good idea

alexandruag commented:
Pushed an extra commit to address comments; I get some weird errors when rebasing on top of latest master so that will need a bit more work.

alexandruag commented:
Pushed an extra commit to address the breakages post compression and noop bump.

danenbm commented on bubblegum/program/tests/simple.rs:
```diff
@@ -54,54 +65,211 @@ async fn test_create_tree_and_mint_passes() {
+        tree.verify_creator(leaf, &context.default_creators[0])
+            .await
+            .unwrap();
+    }
```
Yes I think comments would be great solution.

alexandruag commented:
Rebased again and also added the comments regarding the on-chain tree root and expected value being checked implicitly by the operations based on `TxBuilder::execute` to individual tests in `simple.rs`. As a thought, when breaking tests into multiple files later, we can probably keep the reminder as module/file level documentation.

danenbm commented on bubblegum/program/download-compression-programs.sh:
```diff
@@ -4,19 +4,17 @@
 
 mkdir -p solana_program_library
-curl -LkSs https://api.github.com/repos/solana-labs/solana-program-library/tarball/5f189d1fc0722a619c9b850f2769f85fccb5baea | tar -xz --strip-components=1 -C ./solana_program_library
+curl -LkSs https://api.github.com/repos/solana-labs/solana-program-library/tarball/b0d84fc4 | tar -xz --strip-components=1 -C ./solana_program_library
```
This probably should be updated to 6e8179484413cd8a50ee794c0a49cfef91e1fcb66e8179484413cd8a50ee794c0a49cfef91e1fcb6 or newer

danenbm commented on bubblegum/program/tests/simple.rs:
```diff
@@ -1,7 +1,18 @@
 pub mod utils;
```
It can wait.  Please make sure it is captured in a linear task if you plan to do it later.  Thanks!

danenbm reviewed (approved):
Couple minimal comments.  LGTM!

danenbm reviewed (dismissed):
Couple minimal comments. LGTM!
